Organize Settings (Type & Category)

benjaminjrobin opened the pull request:
Add SettingType and SettingCategory to organize the Settings page. This is entirely data-drive. Here's an example:
![image](https://user-images.githubusercontent.com/14826107/62473385-3e394a80-b766-11e9-8fde-1d0782fe70c0.png)

Also, add Boolean and Date data types for Settings:
<img width="918" alt="Screen Shot 2019-08-05 at 9 49 43 AM" src="https://user-images.githubusercontent.com/14826107/62473444-5e690980-b766-11e9-8278-547c5f848df1.png">

benjaminjrobin commented:
@brondavies It looks like the CodeFactor linter is running on code generated by the `DbmlBuilder` tool that keeps the C# code in sync with the sql migrations. Do we want to disable that for generated code or should I plan to go fix those by hand?

Cc: @drmohundro

brondavies commented:
> @brondavies It looks like the CodeFactor linter is running on code generated by the `DbmlBuilder` tool that keeps the C# code in sync with the sql migrations. Do we want to disable that for generated code or should I plan to go fix those by hand?
> 
> Cc: @drmohundro
As long as that's all it's complaining about, we'll ignore it.  We would like to fix the generated code in the not too distant future

brondavies commented on CmsData/Migrations/20190718_create_setting_type_and_category.sql:
```diff
@@ -0,0 +1,57 @@
+﻿ALTER TABLE [dbo].[Setting]
```
We like to keep these scripts idempotent so that they will run without errors even if they've been run before.  So you'll have to check if the thing your script is going to create/delete is/n't there first.  Look at some of the other migration scripts that do this.

brondavies replied:
@benjaminjrobin Keep in mind that each query is run separately - delimited by the `GO` command

brondavies commented on CmsWeb/Areas/Setup/Views/Setting/Index.cshtml:
```diff
@@ -7,6 +9,19 @@
 @section head{
     <meta name="format-detection" content="telephone=no" />
     @Fingerprint.Css("/Content/touchpoint/lib/bootstrap-editable/css/bootstrap-editable.css")
+    <style>
```
We have a new component/standard for expand/collapse buttons used in this PR https://github.com/bvcms/bvcms/pull/528 on the Roles settings page
![roles-accordian](https://user-images.githubusercontent.com/2912683/61965028-e3993500-af94-11e9-9cf8-4fa4b2903128.gif)

brondavies commented on CmsWeb/Areas/Setup/Views/Setting/Index.cshtml:
```diff
@@ -140,16 +199,55 @@
+                    e.preventDefault();
+                    var tab = $(e.target).attr('href').replace("#", "#tab-");
+                    window.location.hash = tab;
+                    $.cookie('last-settings-tab', tab);
```
instead of cookies, use localstorage - fewer limitations and less network traffic

benjaminjrobin commented:
@brondavies Can you give this another pass? I've updated the things you mentioned, and added the remaining functionality for this feature. Not ready to merge quite yet, but should be very soon - letting the PR app build so we can do UAT.

benjaminjrobin commented:
@brondavies UAT passed. Ready for merge here!

brondavies commented on CmsData/Migrations/20190826_Create_Setting_Type_Category_and_Metadata.sql:
```diff
@@ -0,0 +1,57 @@
+			REFERENCES [dbo].[SettingCategory] ([SettingCategoryId])
+		ALTER TABLE [dbo].[SettingMetadata]  WITH CHECK ADD  CONSTRAINT [FK_SettingMetadata_SettingType] FOREIGN KEY([SettingTypeId])
+			REFERENCES [dbo].[SettingType] ([SettingTypeId])
+        ALTER TABLE [dbo].[SettingMetadata]  WITH CHECK ADD  CONSTRAINT [FK_SettingMetadata_Setting] FOREIGN KEY([SettingId])
```
We don't want this FK because it would require an entry in the Setting table.

brondavies commented on CmsData/Migrations/20190826_Create_Setting_Type_Category_and_Metadata.sql:
```diff
@@ -0,0 +1,57 @@
+        ALTER TABLE [dbo].[SettingMetadata]  WITH CHECK ADD  CONSTRAINT [FK_SettingMetadata_Setting] FOREIGN KEY([SettingId])
+			REFERENCES [dbo].[Setting] ([Id])
+
+        INSERT INTO SettingMetadata (SettingId)
```
I am updating this block now so we have the list of metadata and related settings, types, and categories.  I will update the `SettingMetadata` table to remove the FK on the `Setting` table too.

benjaminjrobin replied:
Good catch. Thanks!
